### PR TITLE
netcat: add new errno detection helpers, switch netcat over to them and implement netcat logging

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: lxd/main_netcat.go [hunk 1]
```diff
@@ -21,27 +24,83 @@ func cmdNetcat(args []string) error {
...
+	// Detect the latter case and create a netcat log
+	var logFile *os.File
+	var err error
+	ctName := ""
+	logPath := ""
+	namestart := strings.LastIndex(args[1], "/lxd_")
+	nameend := strings.LastIndex(args[1], "_rsync_")
+	// If this happens something is seriously wrong.
+	if namestart < 0 || nameend < 0 {
+		return fmt.Errorf("Unpexpected unix domain socket path \"%s\".", args[1])
```
Comment (brauner) [inline lxd/main_netcat.go]: typo

File: lxd/storage_migration.go [hunk 2]
```diff
@@ -41,6 +41,7 @@ func (s rsyncStorageSourceDriver) Snapshots() []container {
 }
 
 func (s rsyncStorageSourceDriver) SendWhileRunning(conn *websocket.Conn, op *operation, bwlimit string) error {
+	ctName, _, _ := containerGetParentAndSnapshotName(s.container.Name())
```
Comment (albertodonato) [inline lxd/storage_migration.go]: Mostly a nitpick but you could pass s.container as parameter, and have containerGetParentAndSnapshotName call call .Name() on it

File: lxd/main_netcat.go [hunk 3]
```diff
@@ -21,27 +24,83 @@ func cmdNetcat(args []string) error {
...
+	// The socket name on LXD instances prior to netcat log file is:
+	// 	/<some-tmp-path/lxd_rsync_<random-suffix>
+	// and on LXD instances with netcat log file:
+	// 	/<some-tmp-path/lxd_<ctName>_rsync_<random-suffix>
+	// Detect the latter case and create a netcat log
+	var logFile *os.File
+	var err error
+	ctName := ""
+	logPath := ""
+	namestart := strings.LastIndex(args[1], "/lxd_")
```
Comment (albertodonato) [inline lxd/main_netcat.go]: I wonder if it wouldn't be simpler to use regexp matching here

File: lxd/storage_migration.go (hunk 2 above)
Comment (brauner) [inline lxd/storage_migration.go]: Yeah, but we have `container` already in the struct so passing it everywhere would be weird. :)

File: lxd/main_netcat.go (hunk 3 above)
Comment (stgraber) [inline lxd/main_netcat.go]: I tend to stay away from regexp unless absolutely needed, they rarely improve code readability and don't have the best of security track records (though I'm not aware of any regexp compiler CVEs for Go). In a compiled and heavily optimized language like Go, they're also unlikely to give you the performance boost that they would in scripting languages.

Comment (albertodonato) [inline lxd/main_netcat.go]: My comment wasn't so much about performance as about redability and simplicity. There's quite a lot of code here to extract a substring from that string.

Maybe even just extracting that logic to a separate function would make the code clearer

**Comment by stgraber** [comment]
Quick recap of IRC discussion:
 - "lxd netcat" isn't API, we don't care about old socket paths as LXD is respawned on upgrades.
 - We can pass more than one argument to "lxd netcat", it just wasn't done due to historical compat with "nc", lets stop doing that and just pass multiple args.
 - There is no real need to use a unix socket and pollute /tmp, an abstract unix socket would work fine there so maybe switch to that too.

File: lxd/main_netcat.go [hunk 4]
```diff
@@ -17,31 +19,55 @@ import (
...
+	if shared.PathExists(logPath) {
+		os.Remove(logPath)
+	}
+
+	logFile, err := os.OpenFile(logPath, os.O_WRONLY|os.O_CREATE|os.O_SYNC, 0644)
+	if err != nil {
+		logFile = nil
+	} else {
+		defer logFile.Close()
+	}
```
Comment (stgraber) [inline lxd/main_netcat.go]: Lets just make logfile creation failure fatal.
If /var/log/lxd/ can't be written to, you have much bigger problems than a failed container migration :)

That would also simplify the rest of this change by not having to constantly check for logFile != nil

Comment (brauner) [inline lxd/main_netcat.go]: done

**Pull request merged**